htsget download endpoints (#695)
neicnordic/sensitive-data-archive: sda-download/api/api.md

# API

The Download API service provides functionality for downloading files from the Archive.
It implements the [Data Out API](https://neic-sda.readthedocs.io/en/latest/dataout/#rest-api-endpoints).

Further, it enables the endpoints `/s3` and `/s3-encrypted`, used for htsget.

> The response can be restricted to only contain a given range of a file, and the files can be returned encrypted or decrypted.

Author: Good point! I updated this in 0414773df3c649b288bb2db14c0bad686d383f28

All endpoints require an `Authorization` header with an access token in the `Bearer` scheme.
```
Authorization: Bearer <token>
```
### Authenticated Session

> The client can establish a session to skip time-costly visa validations for further requests. Session is based on the `SESSION_NAME=sda_session_key` (configurable name) cookie returned by the server, which should be returned in later requests.

Reviewer: ```suggestion
The client can establish a session to bypass time-costly visa validations for further requests. The session is established based on the `SESSION_NAME=sda_session_key` (configurable name) cookie returned by the server, which should be included in later requests.
```

## Service Description

### Endpoints overview:

**Data out API**:

- `/metadata/datasets`
- `/metadata/datasets/*dataset`
- `/files/:fileid`

**htsget**:

- `/s3/<datasetid>/<fileid>`
- `/s3-encrypted/<datasetid>/<fileid>`

### Data out API
#### Datasets

> The `/metadata/datasets` endpoint is used to display the list of datasets the given token is authorised to access, that are present in the archive.

Reviewer: ```suggestion
The `/metadata/datasets` endpoint is used to display the list of datasets that the given token is authorised to access, and that are present in the archive.
```

##### Request
```
GET /metadata/datasets
```
##### Response
```
[
    "dataset_1",
    "dataset_2"
]
```
#### Files
##### Request

> Files contained by a dataset are listed using the `datasetName` from `/metadata/datasets`.

Reviewer: ```suggestion
The files contained in a dataset are listed using the `datasetName` obtained from `/metadata/datasets` endpoint.
```

```
GET /metadata/datasets/{datasetName}/files
```
**Scheme Parameter**

> The `?scheme=` query parameter is optional. When a dataset contains a scheme, it may sometimes be problematic with reverse proxies.

Reviewer: ```suggestion
The `?scheme=` query parameter is optional. When a dataset contains a scheme, it may sometimes encounter issues with reverse proxies.
```

> The scheme can be split from the dataset name, and supplied in a query parameter.

Reviewer: ```suggestion
The scheme can be separated from the dataset name and supplied in a query parameter.
```

##### Response
```
[
    {
        "fileId": "urn:file:1",
        "datasetId": "dataset_1",
        "displayFileName": "file_1.txt.c4gh",
        "fileName": "hash",
        "fileSize": 60,
        "decryptedFileSize": 32,
        "decryptedFileChecksum": "hash",
        "decryptedFileChecksumType": "SHA256",
        "fileStatus": "READY"
    },
    {
        "fileId": "urn:file:2",
        "datasetId": "dataset_1",
        "displayFileName": "file_2.txt.c4gh",
        "fileName": "hash",
        "fileSize": 60,
        "decryptedFileSize": 32,
        "decryptedFileChecksum": "hash",
        "decryptedFileChecksumType": "SHA256",
        "fileStatus": "READY"
    },
]
```
#### File Data
File data is downloaded using the `fileId` from `/metadata/datasets/{datasetName}/files`.
##### Request
```
GET /files/{fileId}
```
##### Response
Response is given as byte stream `application/octet-stream`
```
hello
```
##### Optional Query Parameters
Parts of a file can be requested with specific byte ranges using `startCoordinate` and `endCoordinate` query parameters, e.g.:
```
?startCoordinate=0&endCoordinate=100
```

### S3 requests, for htsget

**Parameters**:

*Partial file retrieval*: 

> - `startCoordinate` Start byte position in the file. If the request is for an encrypted file, the position will be adjusted to align with the nearest data block boundary.

Reviewer: ```suggestion
- `startCoordinate`: start byte position in the file. If the request is for an encrypted file, the position will be adjusted to align with the nearest data block boundary.
```

> - `endCoordinate` End byte position in the file. If the request is for an encrypted file, the position will be adjusted to align with the nearest data block boundary.

Reviewer: ```suggestion
- `endCoordinate`: end byte position in the file. If the request is for an encrypted file, the position will be adjusted to align with the nearest data block boundary.
```

Headers:

- `Authorization: Bearer <token>` 
- `Range: bytes=<start>-<end>`  exact positions. Overrides parameter coordinates.
- `Client-public-key: <key>` used for re-encrypting the header of the file before sending it.

> - `Server-public-key: <key>` used in communication with htsget, for re-encrypting the header of the file.

Reviewer 2: nitpicking 
- can't see to seem to find how are these handled
- do these come in the same request that they need to be separate ?
- there was a diagram how these requests looked like https://github.com/neicnordic/sda-download/blob/fbc45ed5885524032c5b0e0dac859c17a7f1af85/biohackathon/paper.md#htsget a new one how this would look like would help

Reviewer 2: reason I am asking is that:
- i think it can be the same header name if those come in separate requests
- if it is htsget server that includes those headers the request i think we can have that header be an array
- (opinion) in general maybe it should be an array so that we allow multiple keys

Author: Here is the [new diagram](https://raw.githubusercontent.com/umccr/htsget-rs/crypt4gh/docs/crypt4gh/htsget-rs-crypt4gh.drawio.png), for context [see  this](https://github.com/umccr/htsget-rs/blob/crypt4gh/docs/crypt4gh/ARCHITECTURE.md#the-protocol).

Author: We do not use these headers yet, but the plan is to do so as soon as we have the re-encryption server connected (#698).

Reviewer 2: then maybe we can include them with PR for that issue.

Reviewer 2: talked with @dbampalikis and understood why these headers are like this, but still splitting them from this PR makes more sense

Author: Headers are removed in 8cece7d

> # API

Author: Thanks for the review :+1: !
Yes... I thought it would be nice to have it under `api/` for consistency with the services in `sda/cmd`. But you are right that it's better to have things in one place, so I merged the `docs/API.md` with the new one in
c49f140abc9f2597733b53b1bc17cdcd424b108c. Please have look!
